### PR TITLE
Finished

### PR DESCRIPTION
**Pull request by steveryan**
It's all working, but I'm sure it could be made a lot cleaner.

### CODE REVIEW

File: user.rb [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
-  @@all_users = []
+  @all_users = []
   attr_accessor :first_name, :last_name, :email
 
   def self.all
-    @@all_users
+    @all_users
+  end
+
+  def self.all=(array_of_users)
```
Comment (bswinnerton) [inline user.rb:11]: Nice 👍. Rails has the concept of a `cattr_accessor`, which does something similar (but I think ultimately still uses class variables).

File: user.rb [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
+    @all_users = array_of_users
   end
 
   def self.create(first_name:, last_name:, email:)
     if User.is_email_in_use?(email)
-       return puts "Sorry this email is taken"
+       return puts "Sorry, the email #{email} is already in use"
```
Comment (bswinnerton) [inline user.rb]: I think a custom exception would be better suited here. `puts` is usually only used in debugging.

What do you think of something like:

```suggestion
       raise ArgumentError "Sorry, the email #{email} is already in use"
```

File: user.rb [hunk 3]
```diff
@@ -22,20 +26,26 @@ def self.create(first_name:, last_name:, email:)
   end
 
   def self.write_new_user_to_db
```
Comment (bswinnerton) [inline user.rb:28]: I wonder if there's a way that we could abstract away the persistence logic into a separate file (a module), that's not tied to users.

File: user.rb [hunk 4]
```diff
@@ -45,7 +55,7 @@ def initialize(first_name:,
     @first_name = first_name
     @last_name = last_name
     @email = email
-    @@all_users << self
+    User.all=( User.all << self )
     User.write_new_user_to_db
```
Comment (bswinnerton) [inline user.rb:59]: Having a side effect that writes to the file could be a big performance hit. Every time that a new user is instantiated, Ruby needs to open up a file and write to it (which can get quite expensive). Is this necessary here?

File: user.rb (hunk 3 above)
Comment (bswinnerton) [inline user.rb:28]: Would this be better named as `write_all_users_to_db`?

**Review by bswinnerton — APPROVED** [review]
This looks awesome, @steveryan.

Looking at the `update` method, what would happen if a user doesn't pass a `first_name` or `last_name`? I _think_ it would nullify those values.

File: user.rb (hunk 3 above)
Comment (steveryan) [reply user.rb:28]: What persistence logic do you mean? Everything involved in reading/writing to the file?

Comment (steveryan) [reply user.rb:28]: It certainly would. I originally had two separate methods, but collapsed them into one without updating the name to reflect that.

File: user.rb (hunk 4 above)
Comment (steveryan) [reply user.rb:59]: Which place are you referring to? I guess that doesn't matter really. The only reason I write to the file after every single change is that I never know when someone is going to close the program. If I could have a method trigger when the program is closed (or right before its's closed), I could just write to the file once at the very end of a user's session, similar to how I only read from the file once, when the program is opened.

I just don't know of a way to do that. Is there one?

**Comment by steveryan** [comment]
> This looks awesome, @steveryan.
> 
> Looking at the `update` method, what would happen if a user doesn't pass a `first_name` or `last_name`? I _think_ it would nullify those values.

Nope, I'm passing in the users current first, last, and email as the defaults, so if they don't update any (or all) of those fields, they just aren't changed.

**Pull request merged**